test/he/test_pmtud.c

```diff
@@ -0,0 +1,269 @@
+/**
+ * Lightway Core
+ * Copyright (C) 2023 Express VPN International Ltd.
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 2
+ * of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+#include <unity.h>
+
+#include "he.h"
+#include "he_internal.h"
+#include "test_defs.h"
+
+// Unit under test
+#include "pmtud.h"
+
+// Mocks
+#include "mock_conn_internal.h"
+#include "mock_conn.h"
+
+// Helper macros
+#define EXPECT_HE_INTERNAL_SEND_MESSAGE()                                 \
+  do {                                                                    \
+    he_internal_send_message_ExpectAndReturn(&conn, NULL, 0, HE_SUCCESS); \
+    he_internal_send_message_IgnoreArg_message();                         \
+    he_internal_send_message_IgnoreArg_length();                          \
+  } while(0)
+
+static he_conn_t conn = {0};
+
+void setUp(void) {
+}
+
+void tearDown(void) {
+  memset(&conn, 0, sizeof(he_conn_t));
+  call_counter = 0;
+}
+
+void test_he_internal_pmtud_send_probe(void) {
+  conn.state = HE_STATE_ONLINE;
+  conn.pmtud_state = HE_PMTUD_STATE_BASE;
+  conn.pmtud_time_cb = pmtud_time_cb;
+
+  uint16_t probe_mtu = 1212;
+
+  // The expected ping message length should be equal to the length of the of data message of the
+  // given payload size
+  uint16_t expected_length = probe_mtu + sizeof(he_msg_data_t);
+  he_internal_send_message_ExpectAndReturn(&conn, NULL, expected_length, HE_SUCCESS);
+  he_internal_send_message_IgnoreArg_message();
+  he_return_code_t res = he_internal_pmtud_send_probe(&conn, probe_mtu);
+  TEST_ASSERT_EQUAL(HE_SUCCESS, res);
+  TEST_ASSERT_EQUAL(1, call_counter);
+}
+
+void test_he_internal_pmtud_send_probe_nulls(void) {
+  TEST_ASSERT_EQUAL(HE_ERR_NULL_POINTER, he_internal_pmtud_send_probe(NULL, 1350));
+}
+
+void test_he_internal_pmtud_send_probe_invalid_state(void) {
+  TEST_ASSERT_EQUAL(HE_ERR_INVALID_CONN_STATE, he_internal_pmtud_send_probe(&conn, 1350));
```

Comment:
It's not clear why this would result in `HE_ERR_INVALID_CONN_STATE`.  I think it's again because `conn.pmutd_state = 0`.  So it would be good to explicitly set conn.pmtud_state = HE_PMTUD_STATE_DISABLED before the ASSERTS.

Reply:
This is because `conn->state` is not in ONLINE state. It's not about `pmtud_state`.